var/spack/repos/builtin/packages/avizo/package.py

```diff
@@ -0,0 +1,37 @@
+@@ -0,0 +1,36 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+import os
+import tarfile
+
+
+class Avizo(Package):
+    """Avizo is a 3D analysis software for scientific and industrial data.
+    Wherever three-dimensional imaging data sets need to be processed, in
+    materials science, geosciences or engineering applications, Avizo offers
+    abundant state-of-the-art image data processing, exploration and analysis
+    features within an intuitive workflow and easy-to-use graphical user interface."""
+
+    homepage = "https://www.thermofisher.com/sa/en/home/industrial/electron-microscopy/\
+               electron-microscopy-instruments-workflow-solutions/3d-visualization-\
+               analysis-software.html"
+    version('9.7.0', 'ed3947e61a1d17839005c824df975030')
+
+    ## You have to create the tarball for the source code to be processed correctly.
+    def url_for_version(self, version):
+        return "file://{0}/avizo_{1}_linux64_gcc44.tar.gz".format(os.getcwd(), version)
+
+    def install(self, spec, prefix):
+        ver = str(self.version).replace(".", "")
+        os.system('sh Avizo-{0}-Linux64-gcc44.bin --noexec --keep'.format(ver))
+
+        os.chdir(join_path(self.stage.source_path, 'Avizo'))
+        avizo_tar = tarfile.open(name='Avizo-{0}-Linux64-gcc44.tar.bz2'.format(self.version))
```

Comment:
I'm not sure I understand. You have a `.tar.gz` above. You open it, and it contains a script. You run it, then now you have a `.tar.bz2` tarball to extract? What's going on here?

Reply:
Hi again, sorry it has been a while but got a bit busy.

The reason why this seems complicated and why I created a tarball of my own at the first place is:
The main url for avizo is: 
`ftp://ftp.vsg3d.com/private/MASTERS/Avizo/9.7.0/625f2a1f/Avizo-970-Linux64-gcc44.bin`

But when trying to install avizo, spack ulters the url and thus fails!!
```
spack install avizo
==> Installing avizo
==> Searching for binary cache of avizo
==> Warning: No Spack mirrors are currently configured
==> No binary for avizo found: installing from source
==> Fetching ftp://ftp.vsg3d.com/private/MASTERS/Avizo/9.7.0/625f2a1f/Avizo-9.7.0.bin

curl: (78) RETR response: 550
==> Fetching from ftp://ftp.vsg3d.com/private/MASTERS/Avizo/9.7.0/625f2a1f/Avizo-9.7.0.bin failed.
==> Error: FetchError: All fetchers failed for avizo-test-9.7.0-mcyfp4ecd5erlanxnntcf7rlutjlmuuh

/sw/spack/lib/spack/spack/package.py:916, in do_fetch:
        913                raise FetchError("Will not fetch %s" %
        914                                 self.spec.format('$_$@'), ck_msg)
        915
  >>    916        self.stage.create()
        917        self.stage.fetch(mirror_only)
        918        self._fetch_time = time.time() - start_time
        919
```

After some research, I understood that it's the way how Spack fetches urls, and looks like it splits the url in a way that messes up everything for Avizo url.

If you could advise me how to overcome this, it would be great.

Reply:
It looks like the URL for Avizo contains a random hash anyway, so Spack will never be able to guess the correct URL. You can override the URL on a version-specific basis like so:
```python
version('9.7.0', sha256='...', 
        url='ftp://ftp.vsg3d.com/private/MASTERS/Avizo/9.7.0/625f2a1f/Avizo-970-Linux64-gcc44.bin',
        expand=False)
```
The `expand=False` is needed since you aren't downloading a tarball, just a script.

Reply:
seems to work but It's not installing properly. Please hold merging for now